### PR TITLE
RPED tier-based parts dumping

### PR DESCRIPTION
**Pull request by Hatterhat**
## About The Pull Request
image
Dumping out an RPED (by using it inhand - clickdrag dumping unaffected) causes it to calculate and drop the lowest tier of parts. Emptying further can be done by just using it inhand more.

Directly inspired by https://github.com/Citadel-Station-13/Citadel-Station-13/pull/13763.

The implementation probably sucks. I know I was told to use an associated list but then I forgot how to approach that and decided to slap together the current implementation. Sorry.

https://user-images.githubusercontent.com/31829017/210158742-cae83430-bd9f-407f-a0fa-f11975d15d4c.mp4

## Why It's Good For The Game
Decluttering your part exchanger is as easy as pressing Z or whatever you do to dump out a storage item inhand.

## Changelog
:cl:
qol: Rapid Part Exchange Devices and their bluespace variants now automatically drop their current lowest tier of parts, instead of all of their parts.
/:cl:

### CODE REVIEW

**Comment by Serxule** [comment]
tier 110 thousand? holy shit

**Comment by CRITAWAKETS** [comment]
> tier 110 thousand? holy shit
this mfer shoved ratvar's soul into their RPED

**Comment by Hatterhat** [comment]
> tier 110 thousand? holy shit

its not my fault cells have weird get_part_rating calculations

>this mfer shoved ratvar's soul into their RPED

the part exchanger giveth and the part exchanger taketh away

**Comment by jlsnow301** [comment]
though I love this for the regular RPED, don't bpred's already have unlimited storage? kind of senseless for them unless you're just being very ocd

how does it look with cells and regular components? what about things like glass?

**Comment by davethwave** [comment]
> tier 110 thousand? holy shit

Wouldn't that be 48?

**Comment by Hatterhat** [comment]
> though I love this for the regular RPED, don't bpred's already have unlimited storage? kind of senseless for them unless you're just being very ocd
> 
> how does it look with cells and regular components? what about things like glass?

BRPEDs hold up to 400 items/800 storage volume(?) max, so it's not unlimited but if you're printing a goofy ahh amount of parts you might get there

cells are Weird. this whole thing works off of get_part_rating(), which for non-stock-part items (in this case: stacks, e.g. glass, cables) is, by default, 0. for cells, it's calculated by `maxcharge*10 + charge`, so they return weirdly high values, like in the screenshot for tiered readouts being weird.

i'll update the pr body with a clip demonstrating it

File: code/modules/research/stock_parts.dm [hunk 1]
```diff
@@ -227,11 +227,10 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
...
 	/**
-	 * stacks components like cable,glass,plasteel are not component parts hence their get_part_rating() method is undefined and would return undefined values causing errors
-	 * so we assign them an default rating of 1 when the RPED sorts these components
+	 * even though stacks aren't stock parts, get_part_rating() is defined on the item level (see /obj/item/proc/get_part_rating()) and defaults to returning 0.
 	 */
-	var/a_rating = isstack(A) ? 1 : A.get_part_rating()
-	var/b_rating = isstack(B) ? 1 : B.get_part_rating()
+	var/a_rating = A.get_part_rating()
+	var/b_rating = B.get_part_rating()
 	return b_rating - a_rating
```
Comment (jlsnow301) [inline code/modules/research/stock_parts.dm:234]: Purely a nit here, but for thoroughness I'd like it to be firstObj / secondObj or something (no single letter vars)
Also, can we just do `return secondObj.get_part_rating() - firstObj.get_part_rating()` to skip creating a var?

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: code/modules/research/stock_parts.dm (hunk 1 above)
Comment (Hatterhat) [reply code/modules/research/stock_parts.dm:234]: done!

**Pull request merged**